### ISSUE TITLE
Move all tests from test/ subdir to test/subdir in the component being tested

### ISSUE THREAD

**Issue by flammie** [giellalt/template-lang-und#25]
Reorganising followingly:

* [x] `test/X/Y -> src/X/Y/test`
* [x] all make checks should pass
* [x] fix some tests that don't work with `parallel-tests`
* [x] move `test/run-yaml-testcases.sh.in` and `test/run-morph-tester.sh.in` to `giella-core`. \
    This requires that language-specific parts be moved to other files, at least the enabled fst tech identification. It also requires updating the fst detection algorithm - some tests fail now because they can't find the required FST.
* [x] also factor out common program logic from `src/fst/morphology/test/generate-*`

**Comment by snomos** [comment]
There are a number of files still in the old dir structure:

```
lang-est-x-plamk/test/tools/
└── spellcheckers
    ├── Makefile.am
    └── fstbased
        └── desktop
            └── hfst
                └── accept-all-lemmas.sh.in

lang-sma/test/tools/
├── mt
│   └── apertium
│       └── Makefile.am
└── spellcheckers
    ├── Makefile.am
    ├── fstbased
    │   └── desktop
    │       └── hfst
    │           └── accept-all-lemmas.sh.in
    ├── listbased
    │   └── plx
    │       ├── expected.plx
    │       ├── words-accept.exp-out.txt
    │       ├── words-accept.txt
    │       ├── words-reject.exp-out.txt
    │       ├── words-reject.txt
    │       ├── words.regex
    │       └── words.txt
    └── speller-testbed-sma.txt
```

Must these be cleaned manually, or are they already copied to the new location, so that they can just be deleted?

**Comment by flammie** [comment]
Some are language specific tests that need to be ported still, `test/tools/spellcheckers/listbases/plx` for example. Perhaps  `accept-all-lemmas` is something that could be in template too. And for few fiels like `test/tools/spellcheckers/Makefile.am` that are copies of `tools/spellcheckers/test/Makefile.am` can be deleted, not sure why scripts didn't work there.

**Comment by snomos** [comment]
About `listbased/`: since we do not build such spellers anymore, the data does not need to be moved to the new test dir locations. Instead I suggest we move the data to `test/data/`.

I also suggest that we rename `test/data/` to `testdata/`. The idea is that this dir contains testdata used by several other tests, ie is shared among tests in various locations. Test data that is specific to a (set of) test(s) in one dir should stay with the tests.

**Comment by snomos** [comment]
Things failing after the re-org:

- [ ] https://github.com/giellalt/lang-sma/issues/100 (failing only in bash, including CI, working in zsh)

**Comment by Trondtr** [comment]
Things failing after the re-org (I could have made separate issues, but let us start out here):

1. We do not get analysed and generated missing lemmas in our editor automatically, as before. The files  src/fst/morphology/test/generated_missing_adjectives_lemmas.hfst.txt and ...analysed...  exist, but are not shown automatically, as before.

2. yaml testing report were earlier shown on a file-for-fileb basis, this now does not happen, instead, we get an overall summary.

3. We earlier had a shellscript giving yaml tests directly , in e.g. `lang-smn/test/yaml-check.sh` (cf. git), this now does not work ( I tried both to change the path and to move the script to the new test catalogue, nothing helped)

4. `make check` stops hanging after `PASS: tag_test.sh`(tested on vep and smn, both 25 min and counting). This is not a general problem, as fit gets past this point immediately. nob hangs on the next command (adjectives), there is thus sand in this machinery.

**Comment by flammie** [comment]
I think some of the issues highlight that `make check` is not ideal for the purpose it was used, its main purpose in life is just to act as a litmus test for automation and external users that the downloaded version works within specified parametres (hence `make -j && make -j check && make install`). The tooling to support developers (e.g. opening log files in an editor) should probably be decoupled to `devtools/` or `devtest`.

One way to use the current test driver efficiently is to keep `test-suite.log` open in an editor that supports automatic reloading, it should contain most of the stuff previous test driver outputted on terminal (for failing tests).

**Comment by flammie** [comment]
the make check uses now `see &` if giella-core-1.0.1 is configured to find one, but I haven't tested it, on linux etc. it falls back to `head` which is quite fine for logs

**Comment by snomos** [comment]
> About `listbased/`: since we do not build such spellers anymore, the data does not need to be moved to the new test dir locations. Instead I suggest we move the data to `test/data/`.
> 
> I also suggest that we rename `test/data/` to `testdata/`. The idea is that this dir contains testdata used by several other tests, ie is shared among tests in various locations. Test data that is specific to a (set of) test(s) in one dir should stay with the tests.

I have changed my mind about this one: I would like to get rid of the `test/` catalog altogether, as it is confusing that it is still there - what is the difference between that one and the `test/` dirs in various subdirectories?

In the template, the test dir presently contains the following:

```
template-lang-und/test/
├── Makefile.am
├── check_lemmas.sh
├── data
│   ├── analysis_check_corpus.txt
│   └── typos.txt
├── error-handling-stubs.sh
├── run-morph-tester.sh.in
└── run-yaml-testcases.sh.in
```

All shell files should be moved, either to `giella-core` or to a suitable `test/` dir. `typos.txt` should be moved to `tools/spellcheckers/test/`, and the corpus analysis data should be moved to `tools/analysers/test/`.

Additonal data and tests in various languages have to be dealt with on a case by case basis.

@flammie could you have a look? It would be great to close this one soon 🙂

**Comment by flammie** [comment]
Yeah I think template part can be taken down and the language specific testers that still use these scripts will turn out in the process.
